### ISSUE TITLE
Potential saftey issue

### ISSUE THREAD

**Issue by bbatha** [Kimundi/scoped_thread_pool_cache#8]
Using the following code I am able to reliably corrupt data in foo in rust nightly rustc 1.5.0-nightly (b2f379cdc 2015-09-23)

``` Rust

extern crate scoped_threadpool;
use scoped_threadpool::{Pool, Scope};

#[derive(Debug)]
pub struct RefOwner<'a>(&'a str);

pub struct ScopeRef<'a, 'pool, 'scope> where 'pool: 'scope, 'scope: 'a {
    scope: &'a Scope<'pool, 'scope>,
}

impl<'a, 'pool, 'scope> ScopeRef<'a, 'pool, 'scope> {
    pub fn new(scope: &'a Scope<'pool, 'scope>) -> ScopeRef<'a, 'pool, 'scope> {
        ScopeRef { scope: scope }
    }

    pub fn execute<'b: 'pool>(&self, ref_owner: &'b RefOwner) -> String {
        self.scope.execute(move || {
            println!("Got: {}", ref_owner.0);
        });
        ref_owner.0.to_owned()
    }
}

#[test]
fn it_works() {
    let foo = "foo".to_owned();
    let mut pool = Pool::new(2);

    pool.scoped(|scope| {
        let ref_owner= RefOwner(&foo);
        let scope_ref = ScopeRef::new(scope);
        let result = scope_ref.execute(&ref_owner);
        assert_eq!(foo, result);
    });
}
```

```
     Running target/debug/pool_problems-5eb09be270cfb456

running 1 test
Got: H
test it_works ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured

   Doc-tests pool_problems

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured
```

**Comment by bbatha** [comment]
Using `move |scope| { ... }` does not fix the error either and causes the process to crash:

```
running 1 test
Process didn't exit successfully: `/home/user/bbatha/experiments/pool_problems/target/debug/pool_problems-5eb09be270cfb456` (signal: 11)
```

**Comment by bstrie** [comment]
Is this because of some unsafe code in scoped_threadpool, or is this a soundness bug in Rust?

**Comment by bbatha** [comment]
This is a soundness bug in rust see rust-lang/rust#28609

**Comment by bstrie** [comment]
@bbatha 28609 appears to be closed, does this still reproduce for you? We should be able to either close this bug or reopen that one.

**Comment by bbatha** [comment]
Looks fixed in 1.4 through nightly.
